tests/unit/TestNetRemoteDataStreamingServiceClient.cxx

```diff
@@ -0,0 +1,136 @@
+            StartCall();
+            NextWrite();
+        }
+
+        void
+        OnWriteDone(bool isOk) override
+        {
+            if (isOk) {
+                NextWrite();
+            } else {
+                StartWritesDone();
+            }
+        }
+
+        void
+        OnDone(const grpc::Status& status) override
+        {
+            std::unique_lock lock(m_mutex);
+
+            m_status = status;
+            m_done = true;
+            m_cv.notify_one();
+        }
+
+        grpc::Status
+        Await(WifiDataStreamUploadResult* result)
+        {
+            std::unique_lock lock(m_mutex);
+
+            m_cv.wait(lock, [this] {
+                return m_done;
+            });
+            *result = m_result;
+
+            return m_status;
+        }
+
+    private:
+        void
+        NextWrite()
+        {
+            if (m_numberOfDataBlocksToWrite > 0) {
+                m_data.set_data(std::format("Data #{}", ++m_numberOfDataBlocksWritten));
+                StartWrite(&m_data);
+                m_numberOfDataBlocksToWrite--;
+            } else {
+                StartWritesDone();
+            }
+        }
+
+    private:
+        grpc::ClientContext m_clientContext{};
+        WifiDataStreamUploadData m_data{};
+        WifiDataStreamUploadResult m_result{};
+        uint32_t m_numberOfDataBlocksToWrite{};
+        uint32_t m_numberOfDataBlocksWritten{};
+        grpc::Status m_status{};
+        std::mutex m_mutex{};
+        std::condition_variable m_cv{};
+        bool m_done{ false };
+    };
+
+    SECTION("Can be called with one client")
+    {
+        static constexpr auto numberOfDataBlocksToWrite = 10;
+        auto dataStreamWriter = std::make_unique<DataStreamWriter>(client.get(), numberOfDataBlocksToWrite);
+
+        WifiDataStreamUploadResult result{};
+        grpc::Status status = dataStreamWriter->Await(&result);
+        REQUIRE(status.ok());
+        REQUIRE(result.numberofdatablocksreceived() == numberOfDataBlocksToWrite);
+        REQUIRE(result.status().code() == WifiDataStreamOperationStatusCodeSucceeded);
+    }
+
+    SECTION("Can be called with multiple parallel clients")
+    {
+        static constexpr auto numberOfDataBlocksToWrite = 10;
+        static constexpr auto numberOfClients = 5;
+
+        std::vector<std::unique_ptr<NetRemoteDataStreaming::Stub>> dataStreamingClients;
+        std::vector<std::unique_ptr<DataStreamWriter>> dataStreamWriters;
+        std::vector<std::jthread> clientThreads;
+
+        for (std::size_t i = 0; i < numberOfClients; i++) {
+            dataStreamingClients.push_back(NetRemoteDataStreaming::NewStub(channel));
+            dataStreamWriters.push_back(std::make_unique<DataStreamWriter>(dataStreamingClients.back().get(), numberOfDataBlocksToWrite));
+
+            clientThreads.emplace_back([&, i]() {
+                WifiDataStreamUploadResult result{};
+                grpc::Status status = dataStreamWriters[i]->Await(&result);
```

Comment:
You can introduce a new variable in the capture here to make accessing the stream writer a bit simpler. Also, it's safer to move the data accessed by the thread into the thread func itself if it's not shared (which it appears not to be IIUC), effectively transferring ownership of it and then you don't need to worry about its lifetime. At that point, it doesn't seem like the `dataStreamWriters` `vector` is needed at all.

```suggestion
            auto dataStreamWriter = std::make_unique<DataStreamWriter>(dataStreamingClients.back().get(), numberOfDataBlocksToWrite);
            clientThreads.emplace_back([dataStreamWriter = std::move(dataStreamWriter)]() {
                WifiDataStreamUploadResult result{};
                grpc::Status status = dataStreamWriter->Await(&result);
```

You could take that even further and only pass a ref to the `channel` into the capture and create both the client and writer inside the thread func.

Reply:
Good idea, I was having lifetime management issues which led to me storing the clients and the stream writers into vectors, but I like the idea of moving the data into the lambda instead. I'll try it out